File: app/src/test/java/org/metplus/curriculum/security/services/LocalTokenServiceTest.java
package org.metplus.curriculum.security.services;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.metplus.curriculum.test.BeforeAfterInterface;
import org.metplus.curriculum.test.BeforeAfterRule;

import java.time.Clock;
import java.time.Instant;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.when;


@RunWith(Suite.class)
@Suite.SuiteClasses({LocalTokenServiceTest.ValidToken.class,
        LocalTokenServiceTest.GenerateToken.class})
public class LocalTokenServiceTest {

    public static class DefaultLocalTokenServiceTest implements BeforeAfterInterface {
        Clock clock;
        Instant now;
        Instant afterTimeout;
        @Rule
        public BeforeAfterRule beforeAfter = new BeforeAfterRule(this);
        protected LocalTokenService service;
        protected String generatedKey;

        @Override
        public void after() {
        }

        @Override
        public void before() {
            clock = mock(Clock.class);
            now = Instant.now();
            afterTimeout = now.plusSeconds(1800);
            when(clock.instant()).thenReturn(now);

            service = new LocalTokenService(clock, 1800);
            generatedKey = service.generateToken("1.1.1.1");
        }
    }

    public static class ValidToken extends DefaultLocalTokenServiceTest {
        @Test
        public void validToken_shouldReturnTrue() throws Exception {
            assertTrue(service.isValid(generatedKey));
        }

        @Test
        public void invalidToken_shouldReturnFalse() throws Exception {
            assertFalse(service.isValid("9367f48e-94d3-4b3b-8cbf-55cf61e58fe2"));
        }

        @Test
        public void nullToken_shouldReturnFalse() throws Exception {
            assertFalse(service.isValid(null));
        }

        @Test
        public void tokenNotUUID_shouldReturnFalse() throws Exception {
            assertFalse(service.isValid("123"));
        }

        @Test
        public void timeoutReached_shouldReturnFalse() throws Exception {
            reset(clock);
            when(clock.instant()).thenReturn(afterTimeout);
            assertFalse(service.isValid(generatedKey));
        }

        @Test
        public void timeoutNotReached_shouldReturnTrue() throws Exception {
            reset(clock);
            when(clock.instant()).thenReturn(afterTimeout.minusSeconds(1));
            assertTrue(service.isValid(generatedKey));
        }
    }

    public static class GenerateToken extends DefaultLocalTokenServiceTest {
        @Test
        public void tokenGenerated_shouldBeValid() throws Exception {
            String token = service.generateToken("123.123.123.123");
            assertTrue(service.isValid(token));
        }

        @Test
        public void tokenGeneratedTwiceForIPAddress_shouldReturnSameKey() throws Exception {
            String token = service.generateToken("2.2.2.2");
            assertEquals(token, service.generateToken("2.2.2.2"));
        }

        @Test
        public void tokenGeneratedTwiceForIPAddressAfterTimeout_shouldReturnDifferentKey() throws Exception {
            service = new LocalTokenService(0);
            String token = service.generateToken("2.2.2.2");
            assertNotEquals(token, service.generateToken("2.2.2.2"));
        }

        @Test
        public void tokenGeneratedTwiceForIPAddressAfterTimeout_shouldNotIncreaseTheNumberOfTokens() throws Exception {
            service = new LocalTokenService(0);
            String token = service.generateToken("2.2.2.2");
            int numberOfToken = service.totalNumberTokens();
            service.generateToken("2.2.2.2");
            assertEquals(numberOfToken, service.totalNumberTokens());
        }
    }
}

File: web/src/test/kotlin/org/metplus/cruncher/web/security/services/LocalTokenServiceTest.kt
package org.metplus.cruncher.web.security.services

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.Mockito.reset
import java.time.Clock
import java.time.Instant


open class LocalTokenServiceTestDefaultLocalTokenServiceTest {
    internal lateinit var clock: Clock
    internal lateinit var now: Instant
    internal lateinit var afterTimeout: Instant
    protected lateinit var service: LocalTokenService
    protected lateinit var generatedKey: String

    @BeforeEach
    fun before() {
        clock = mock(Clock::class.java)
        now = Instant.now()
        afterTimeout = now.plusSeconds(1800)
        `when`(clock.instant()).thenReturn(now)

        service = LocalTokenService(clock, 1800)
        generatedKey = service.generateToken("1.1.1.1")
    }
}

class LocalTokenServiceTestValidToken : LocalTokenServiceTestDefaultLocalTokenServiceTest() {
    @Test
    fun validToken_shouldReturnTrue() {
        assertThat(service.isValid(generatedKey)).isTrue()
    }

    @Test
    fun invalidToken_shouldReturnFalse() {
        assertThat(service.isValid("9367f48e-94d3-4b3b-8cbf-55cf61e58fe2")).isFalse()
    }

    @Test
    fun nullToken_shouldReturnFalse() {
        assertThat(service.isValid(null)).isFalse()
    }

    @Test
    fun tokenNotUUID_shouldReturnFalse() {
        assertThat(service.isValid("123")).isFalse()
    }

    @Test
    fun timeoutReached_shouldReturnFalse() {
        reset(clock)
        `when`(clock.instant()).thenReturn(afterTimeout)
        assertThat(service.isValid(generatedKey)).isFalse()
    }

    @Test
    @Throws(Exception::class)
    fun timeoutNotReached_shouldReturnTrue() {
        reset(clock)
        `when`(clock.instant()).thenReturn(afterTimeout.minusSeconds(1))
        assertThat(service.isValid(generatedKey)).isTrue()
    }
}

class LocalTokenServiceTestGenerateToken : LocalTokenServiceTestDefaultLocalTokenServiceTest() {
    @Test
    @Throws(Exception::class)
    fun tokenGenerated_shouldBeValid() {
        val token = service.generateToken("123.123.123.123")
        assertThat(service.isValid(token)).isTrue()
    }

    @Test
    @Throws(Exception::class)
    fun tokenGeneratedTwiceForIPAddress_shouldReturnSameKey() {
        val token = service.generateToken("2.2.2.2")
        assertThat(service.generateToken("2.2.2.2")).isEqualTo(token)
    }

    @Test
    @Throws(Exception::class)
    fun tokenGeneratedTwiceForIPAddressAfterTimeout_shouldReturnDifferentKey() {
        service = LocalTokenService(0)
        val token = service.generateToken("2.2.2.2")
        assertThat(service.generateToken("2.2.2.2")).isNotEqualTo(token)
    }

    @Test
    @Throws(Exception::class)
    fun tokenGeneratedTwiceForIPAddressAfterTimeout_shouldNotIncreaseTheNumberOfTokens() {
        service = LocalTokenService(0)
        val token = service.generateToken("2.2.2.2")
        val numberOfToken = service.totalNumberTokens()
        service.generateToken("2.2.2.2")
        assertThat(service.totalNumberTokens()).isEqualTo(numberOfToken)
    }
}